benchmarks/persistence/persistence.py

```diff
@@ -0,0 +1,232 @@
+
+        Args:
+            strategy: The strategy to construct the path.
+
+        Returns:
+            The path to persist the object. Can be a file name or
+            a folder path as a string.
+        """
+        separator = "/" if strategy is PathStrategy.HIERARCHICAL else "_"
+
+        file_usable_date = self.execution_date_time.strftime("%Y-%m-%d")
+        components = [
+            self.benchmark_name,
+            self.branch,
+            self.latest_baybe_tag,
+            file_usable_date,
+            self.commit_hash,
+        ]
+
+        sanitized_components = [
+            self._string_sanitizer(component) for component in components
+        ]
+        path = separator.join(sanitized_components) + separator + "result.json"
+
+        return path
+
+
+class ObjectWriter(Protocol):
+    """Interface for interacting with storage."""
+
+    def write_json(self, object: dict, path_constructor: PathConstructor) -> None:
+        """Store a JSON serializable dict according to the path.
+
+        If the respective object exists, it will be overwritten.
+
+        Args:
+            object: The object to be persisted.
+            path_constructor: The PathConstructor to create the path for the object.
+        """
+
+
+@define
+class S3ObjectWriter(ObjectWriter):
+    """Class for persisting objects in an S3 bucket."""
+
+    _bucket_name: str = field(validator=instance_of(str), init=False)
+    """The name of the S3 bucket where the results are stored."""
+    _object_session: Session = field(factory=boto3.session.Session)
+    """The boto3 session object. This will load the respective credentials
+    from the environment variables within the container."""
+
+    @_bucket_name.default
+    def _default_bucket_name(self) -> str:
+        """Get the bucket name from the environment variables."""
+        if not PERSIST_DATA_TO_S3_BUCKET:
+            raise ValueError(
+                "No S3 bucket name provided. Please provide the "
+                + "bucket name by setting the environment variable "
+                + "BAYBE_BENCHMARKING_PERSISTENCE_PATH."
+            )
+        return os.environ["BAYBE_BENCHMARKING_PERSISTENCE_PATH"]
+
+    @override
+    def write_json(self, object: dict, path_constructor: PathConstructor) -> None:
+        """Store a JSON serializable dict according to the path.
+
+        This method will store an JSON serializable dict in an S3 bucket.
+        The S3-key of the Java Script Notation Object will be created by
+        the path object. I the key already exists, it will be overwritten.
+
+        Args:
+            object: The object to be persisted.
+            path_constructor: The PathConstructor to create the path for the object.
+
+        """
+        client = self._object_session.client("s3")
+
+        client.put_object(
+            Bucket=self._bucket_name,
+            Key=path_constructor.get_path(strategy=PathStrategy.HIERARCHICAL),
```

Comment:
I'm confused - why does this use HIERARCHICAL? Doesn't this mean that folders are being created, and the whole point of S3 problems were that there are no folder structures there?

Reply:
The syntax of S3 keys resembles a folder structure, which is why the "folderish" strategy is employed. However, S3 does not create actual folders, even though it may appear that way.

Reply:
Ahhh I see, thanks for clarifying